packages/library-sync/src/commonMain/kotlin/io/realm/kotlin/mongodb/exceptions/SyncExceptions.kt

```diff
@@ -44,22 +51,56 @@ public open class SyncException : AppException {
  *
  * @see io.realm.kotlin.mongodb.sync.SyncConfiguration.Builder.errorHandler
  */
-public class UnrecoverableSyncException : SyncException {
-    internal constructor(message: String) : super(message)
-}
+public class UnrecoverableSyncException internal constructor(message: String) :
+    SyncException(message)
 
 /**
  * Thrown when the type of sync used by the server does not match the one used by the client, i.e.
  * the server and client disagrees whether to use Partition-based or Flexible Sync.
  */
-public class WrongSyncTypeException : SyncException {
-    internal constructor(message: String) : super(message)
-}
+public class WrongSyncTypeException internal constructor(message: String) : SyncException(message)
 
 /**
  * Thrown when the server does not support one or more of the queries defined in the
  * [io.realm.kotlin.mongodb.sync.SubscriptionSet].
  */
-public class BadFlexibleSyncQueryException : SyncException {
-    internal constructor(message: String) : super(message)
+public class BadFlexibleSyncQueryException internal constructor(message: String) :
+    SyncException(message)
+
+/**
+ * Thrown when the server undoes one or more client writes. Details on undone writes can be found in
+ * [compensatingWrites].
+ */
+public class CompensatingWriteException internal constructor(
+    message: String,
+    compensatingWrites: Array<CoreCompensatingWriteInfo>
+) : SyncException(message) {
+    public val compensatingWrites: List<CompensatingWriteInfo> = compensatingWrites.map {
+        CompensatingWriteInfo(
+            reason = it.reason,
+            objectName = it.objectName,
+            primaryKey = it.primaryKey.asRealmAny(),
+        )
+    }
+
+    private fun RealmValue.asRealmAny(): RealmAny? = when (val type = getType()) {
+        ValueType.RLM_TYPE_NULL -> null
+        ValueType.RLM_TYPE_INT -> RealmAny.create(getLong())
+        ValueType.RLM_TYPE_BOOL -> RealmAny.create(getBoolean())
+        ValueType.RLM_TYPE_STRING -> RealmAny.create(getString())
+        ValueType.RLM_TYPE_BINARY -> RealmAny.create(getByteArray())
+        ValueType.RLM_TYPE_TIMESTAMP -> RealmAny.create(RealmInstantImpl(getTimestamp()))
+        ValueType.RLM_TYPE_FLOAT -> RealmAny.create(getFloat())
+        ValueType.RLM_TYPE_DOUBLE -> RealmAny.create(getDouble())
+        ValueType.RLM_TYPE_DECIMAL128 -> RealmAny.create(realmValueToDecimal128(this))
+        ValueType.RLM_TYPE_OBJECT_ID -> RealmAny.create(BsonObjectId(getObjectIdBytes()))
+        ValueType.RLM_TYPE_UUID -> RealmAny.create(RealmUUIDImpl(getUUIDBytes()))
+        else -> throw IllegalArgumentException("Unsupported primary key type: ${type.name}")
```

Comment:
I would be slightly hesitant about throwing here. We don't have exhaustive tests for this (for obvious reasons), so if we at any point make a mistake here, we will accidentally crash apps. I would probably return something like `RealmAny.create("Unknown")` instead similar to what we do with errors we don't know exactly what is.